src/routing.rs

```diff
@@ -0,0 +1,116 @@
+// Logic for selecting the request backend
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+use super::Backend;
+
+lazy_static! {
+    // Round robin counter hashmap, shared by all threads
+    static ref RR_COUNTER: Mutex<HashMap<String, u64>> = {
+        Mutex::new(HashMap::new())
+    };
+}
+
+pub fn router(backends: &[Backend], host_header: &str) -> Option<String> {
+    // Matches a given host header with a backend
+    // Performs load balancing when configured
+
+    let backend = match_backend(backends, host_header)?;
+
+    // Check if load balancing is enabled
+    if backend.backend_type.is_some() && backend.backend_type.clone()? == "loadbalanced" {
+        if backend.locations.is_some() {
+            round_robin_select(
+                backend.name.clone().unwrap(),
+                backend.locations.as_ref().unwrap(),
+            )
+        } else {
+            // Config not valid
+            None
+        }
+    } else if backend.location.is_some() {
+        // Load balancing not enabled, return the single location / backend
+        backend.location.clone()
+    } else {
+        // Config not valid
+        None
+    }
+}
+
+fn match_backend<'a>(backends: &'a [Backend], host_header: &str) -> Option<&'a Backend> {
+    backends
+        .iter()
+        .find(|&backend| backend.name.is_some() && host_header == backend.name.clone().unwrap())
+}
+
+fn round_robin_select(backend_name: String, backends: &Vec<String>) -> Option<String> {
+    // Uses the rr_counter to keep an index of the next backend to select.
+    // This counter is incremented or reset to zero when exceeding the number of backends
+    let backend_count = backends.len();
+    let mut rr_map = RR_COUNTER.lock().unwrap();
```

Comment:
You don't need global counter at all, also this approach can be incorrect under some edge cases.

Would be better to add atomic counter into `Backend` struct to track number of requests.